### ISSUE TITLE
Accessing Mix tasks from release

### ISSUE THREAD

**Issue by jihaia** [bitwalker/exrm#67]
Sorry for the noob question - after I have packaged the release and deployed to new environment, I need to run some tasks in order to initialize the database etc. Are those MIX tasks accessible? is that the intent etc?

**Comment by hahuang65** [comment]
Stuff like `mix ecto.migrate` is pretty tough to do without access to mix tasks here. I've had to copy over my entire source folder to the server in order to run it.

**Comment by shankardevy** [comment]
+1 @hahuang65 . I am also in the same boat trying to execute ecto migrations and found no good solution than to dump my db locally and import it on the server. But may be, it's not what exrm is intended to do?

@bitwalker , is it ok to compare exrm to capistrano or ruby/rails community? 

A capistrano task copies all the source files to server, runs the migration and starts the server (of course, through individual plugins for different tasks).

If exrm can be compared to capistrano, then it's expected that exrm gives a full solution (at least in future).

If exrm is not a deployment tool like capistrano, if it never was intended to solve this problem, then a new project needs to be created adding exrm as its dependency and continue the work where exrm left. Like, adding tasks to be run on server, running migrations, starting the server etc.

**Comment by ericmj** [comment]
Ecto should provide an easy to run migrations without mix tasks. It is possible today with `Ecto.Migrator` (please check the docs) but it should be as easy as running the mix task. Please open an issue on Ecto.

**Comment by josevalim** [comment]
@ericmj do you have any thoughts in mind? To me the plan is to exactly call `Ecto.Migrator`, it is public API and, if you are building releases, you can afford to call `Ecto.Migrator` with your options as part of your upgrade recipe.

**Comment by josevalim** [comment]
Maybe this is a matter of documentation but Ecto already does what needs to be done in order to support releases which is by providing a first-class `Ecto.Migrator`.

**Comment by bitwalker** [comment]
@josevalim Yeah I think this is one of the areas where custom appups are intended to be used, though that assumes you are using hot upgrades, rather than rolling restarts. Do you have any suggestions on a good general way of handling Ecto migrations?

@shankardevy exrm is definitely not trying to be the capistrano of Elixir, it's role is as a release packaging tool only - orchestrating deployment should be handled by tools appropriate to the user's given environment and needs, whether that's something like capistrano for Elixir, hand rolled scripts, or something else really varies case by case I think.

**Comment by ericmj** [comment]
@josevalim The `Ecto.Migrator` is already quite good actually, better than I thought. One possible improvement is to remove the `directory` option since that is configurable anyway.

**Comment by HashNuke** [comment]
Just realized that it's not possible to run migrations in a Phoenix app, when the app is running from a release.

My hackish way right now: pass an env var to turn on `server: true` in prod config when running `mix release`. This way other mix tasks can be used.

cc: @chrismccord

**Comment by alexjp** [comment]
@HashNuke I am currently using exrm, where i can successfully and easily deploy a release with it on various machines. Only problem is migrations. Could you share your method ? As if I was a complete beginner with only a few months of elixir ? :)

**Comment by bitwalker** [comment]
@HashNuke @alexjp If you have a solution that works, I'd like to document it and add it to exrm's docs :)

**Comment by HashNuke** [comment]
@bitwalker @alexjp Drafted a blog post yesterday about this and another thing (easier versioning). I'll link you to the blog post once I finish it.

The code I've been using to test is online. Here's how:
* Add this block of code to your `config/config.exs`: https://github.com/HashNuke/phoenix-test/blob/master/config/config.exs#L25-L28
* Now to generate a release run `SERVER=1 mix release`
* When running mix tasks just run `mix foo.bar`

The `SERVER=1` makes all the difference. The value doesn't have to be `1`. Just set the `SERVER` env var to anything.

**Comment by josevalim** [comment]
If we are thinking about the problem generally, the biggest issue is that the scripting mode of mix does not really map to how we run apps in production with releases. For example, to run a task, we don't need to preload all modules as part of the release. 

That's not to say there isn't a solution, quite the opposite, there are many ways to solve this:

1. If you are building your release in the production machine, you can do things like: `mix do release, ecto.migrate && run_release`

2. Allow options to be passed to the release executable, so we can do things like `./run_release --app phoenix:serve_endpoints true` or pass extra config files `./run_release --config server` and `./run_release --config migrator`.

3. Instead of running migrations in a separate node, you can just connect to a random node and execute the migrator thing.

4. Ship with mix tasks inside the release and provide an easy way to run them via exrm. This is a bit cumbersome though because we likely want to use very few tasks and not all of them

I am not sure which one would be the best way to go because it depends on your workflow. If your plan is to build releases locally and just run them in production then we need to choose something like 2 or 3

Also, exrm as far as I know is not a replacement for capistrano, because it is about how to build your system, and not how to deploy it. Although bringing more deploy related features would likely streamline the process a lot (and help solve this problem). For example, if we pick 1 as our default workflow, it is easy to pass tasks to run before executing the release. @MSch may have some input on this.

**Comment by alexjp** [comment]
As i understand that this might be "bloating" exrm, solution 2 is exactly what i need.
It would be great to have migrations in the release, because from experience, when you separate db maintenance from code, sooner or later all hell breaks loose. (having db not in sync with what code expects it to be)

**Comment by HashNuke** [comment]
@josevalim option-2 feels more friendlier. Being able to pass options to the release would be a huge goodie bag.

**Comment by bitwalker** [comment]
@josevalim In Option 2, what is handling the options being sent to the executable? I can modify the boot script to pass options appended to `-extra` (currently `--no-halt +iex` is passed along with the `console` task, and `foreground` passes along any extra arguments to the release unaltered. In the case of the `console` task, `IEx.CLI` gets those options, but in the case of `start` or `foreground`, I'm not sure what happens to them, I assume they are accessible somehow, but I haven't poked around in that area at all.

**Comment by josevalim** [comment]
@bitwalker it depends. If those configs are erlang configs, I assume you would also be able to pass them to the erl `-config` flag? It is worth mentioning that the options I mentioned above do not consider the exrm side (i.e. I am not sure how hard they are or if they are even possible to implement).

**Comment by olivermt** [comment]
I have solved this by just running mix ecto.migrate in my aws codedeploy scripts (since its indempotent anyways).

Is a writeup on that something that you guys would be interested in having as companion-documentation to this repository?

**Comment by bitwalker** [comment]
@olivermt Definitely :)

**Comment by olivermt** [comment]
Allright, will send pull request either next week or week after, a bit clogged up right now :)

**Comment by gdon** [comment]
Hi! Is there some plan to enable the Ecto migrate (and create) through  run-app scritps in production?
As the sugestion number 2 from josevalim.
This would be very useful.
And thanks for the wonderful EXRM!

**Comment by bitwalker** [comment]
@gdon Currently there is no plans to support running Mix tasks with releases. My recommendation is to use the Ecto.Migrator API directly, and use configuration flags to determine whether or not to run the migrator. In my own apps, I run migrations every time the app boots, since they are idempotent.

**Comment by gdon** [comment]
Ok @bitwalker .
Thanks for adivise.

**Comment by iwarshak** [comment]
@bitwalker can you show how you are calling Ecto.Migrator?

**Comment by zmoshansky** [comment]
> @bitwalker can you show how you are calling Ecto.Migrator?
+1, I would also be very interested to see how Ecto migrations can be performed. Especially as part of hot upgrades.

**Comment by schurig** [comment]
> @bitwalker can you show how you are calling Ecto.Migrator?

+1

**Comment by schurig** [comment]
@zmoshansky @iwarshak @bitwalker 

I found a way to run migrations when the server boots up: 

`# File: /lib/your_app.ex`

```
Supervisor.start_link(children, opts)
```

becomes

```
res = Supervisor.start_link(children, opts)

path_to_migrations =
  case Mix.env do
    :dev  -> "priv/repo/migrations"
    _     -> "/home/apps/your-app/current/priv/repo/migrations"
  end
IO.puts "######### run migrations..."
Ecto.Migrator.run(YourApp.Repo, path_to_migrations, :up, all: true)

res
```

This doesn't feel right but it's the best approach I came up with so far - at least without having to include mix tasks in my release. Are there any better approaches?

**Comment by bitwalker** [comment]
@schurig That's exactly the right approach, but here's what I would change:

- Reference `priv/repo/migrations` with `Path.join(["#{:code.priv_dir(:yourapp)}", "repo", "migrations"])`. This will work in all environments.
- Running your migrations every time is fine, but you could also take the approach of passing a flag or setting an environment variable if you only want to run them conditionally. I prefer the approach of running them every time myself, but just so you are aware that there are some other options.
- You should be running your migrations *before* starting your supervisor tree (if the migrations fail, your application shouldn't start).

**Comment by schurig** [comment]
@bitwalker thank you!

I've tried to run my migrations before starting the supervisor tree but then it complains that the MyApp.Repo isn't started:

` ** (ArgumentError) repo MyApp.Repo is not started, please ensure it is part of your supervision tree`


Current version:

```elixir
res = Supervisor.start_link(children, opts)
IO.puts "######### run migrations..."
Ecto.Migrator.run(WorkshopsDe.Repo, Path.join(["#{:code.priv_dir(:workshops_de)}", "repo", "migrations"]), :up, all: true)
res
```

**Comment by bitwalker** [comment]
@schurig Ah right, forgot about that. With that in mind, what you have is probably the best solution I can think of at the moment. If you end up finding a better approach, please let me know so I can document it!

**Comment by MSch** [comment]
Now that https://github.com/bitwalker/exrm/issues/286 is fixed I would recommend the following approach, which is how we're doing it:

Build an `escript` which invokes the tasks you want available in the release (using an exrm plugin: https://gist.github.com/MSch/9cfc185e24e2f3509650)  and invoke it from the release:

```
rel/pssync/bin/pssync escript bin/release_tasks.escript migrate
```

This way we can control when and where to migrate, seed, etc.

There is some duplication/indirection since we want (most) of the tasks available both via `mix` in development and `release_tasks.escript` in production, but that is manageable. It also means we can migrate without fully starting our OTP application.

In an ideal future this would be upstreamed into exrm and more tightly integrated so that I can just do `rel/pssync/bin/pssyc migrate` but so far we've already had a great experience with our approach.

**Comment by beardedprojamz** [comment]
@MSch Thanks for the comment and gist! Worked great for me - easy to expand for my purposes as well.

**Comment by bitwalker** [comment]
I'll get this added to the documentation soon so it's easier to find.

**Comment by killerswan** [comment]
Oooh, @MSch, thanks!

**Comment by MSch** [comment]
FYI @codermctwist @edgurgel @asaaki @killerswan We're now moving to `bin/pssync command PS.ReleaseTasks migrate` since https://github.com/bitwalker/exrm/pull/295 and https://github.com/bitwalker/exrm/pull/319 have been merged. You need exrm 1.0.3 for this.

**Comment by olivermt** [comment]
@MSch could you post a code example of how you do migrations like that?

**Comment by martin-langhoff** [comment]
@MSch , @bitwalker so the recipe is to

* pull out the logic out from the Mix Tasks to a module that does not depend on Mix
* In dev, have `Mix.Task.Foo` call `MyApp.Task.Foo`
* In prod, `myapp.sh command MyApp.Task Foo`

I like the approach, +1 to having it documented it somewhere :-)

**Comment by martin-langhoff** [comment]
It has been a dog to make this work, as any error results in uninformative failure in do_boot.

In the end, the right invocation ended up requiring an Elixir prefix, so on the command line I call:

```
rel/bin/myapp command Elixir.MyApp.DoSomething run
```

**Comment by martin-langhoff** [comment]
Hmm, any functions needing parameters fail. In this context, how do we get an args array built?

**Comment by jwarlander** [comment]
I couldn't figure out parameter passing either; I just arranged it so that I have a function for each purpose; eg:

    bin/wolf command Elixir.Wolf.ReleaseTasks create
    bin/wolf command Elixir.Wolf.ReleaseTasks migrate
    bin/wolf command Elixir.Wolf.ReleaseTasks seed

My `Wolf.ReleaseTasks` module is directly modeled after the one @MSch posted about (https://gist.github.com/MSch/9cfc185e24e2f3509650), and can be found here:

https://gist.github.com/jwarlander/809deb2bb06c2c43abafd471591b2dea

The differences:
- No `main` function, as I couldn't get parameters to work
- Added `create` / `drop` to support initial deployment etc
- Ended up having to call `System.halt(0)` at the end of each exposed function, or the command wouldn't finish

It feels like much of this could be made to be generic.

**Comment by martin-langhoff** [comment]
My experience (and practice) matches that of @jwarlander ; however for some commands/actions I need to be able to pass parameters.

**Comment by MSch** [comment]
@martin-langhoff @jwarlander Look at the example code in the PR for how parameter passing works: https://github.com/bitwalker/exrm/pull/295/files

`:init.get_plain_arguments()` returns the arguments. I agree, it's rather annoying. also the results are Erlang charlists, so you have to do `Enum.map(:init.get_plain_arguments(), &List.to_string/1)`

One more thing, because `bin/pssync command Elixir.PS.ReleaseTasks` is really annoying we gave the module a non-Elixir namespaced name, that works this:

```elixir
defmodule :sync_release do
end
```

Now we can do `bin/pssync command sync_release migrate`

**Comment by MSch** [comment]
> It feels like much of this could be made to be generic.

I would <3 a way to register commands and then being able to do `bin/<release_name> <command_name> <args>` instead of having to do `bin/<release_name> command <module> <function> <args>`. But for us it's a low priority thing.

> Ended up having to call System.halt(0) at the end of each exposed function, or the command wouldn't finish

Better do `:init.stop(0)`.

**Comment by martin-langhoff** [comment]
For the record, I have used @MSch 's advice and it's worked beautifully.

there are some issues still on v1.0.3 around boot_clean vs boot that I cannot quite figure out.

**Comment by bitwalker** [comment]
Just to follow up regarding #340, I pushed a fix in 1.0.5 for the `console_clean` command.

**Comment by martin-langhoff** [comment]
Yes I can confirm that on 1.0.5 this works great. Thank you!

One thing I have ended up needing for mnesia DB prep and seeding is a function that mimics what Application.ensure_all_started() does, except that it only starts the deps, not the app itself.

**Comment by martin-langhoff** [comment]
So I declare
```
  @shortdoc "Start all deps without starting the app itself"
  def ensure_all_deps_started() do
    Application.load(:foo)
    Enum.map(Application.spec(:foo, :applications), &Application.load(&1))
    Enum.map(Application.spec(:foo, :applications), &Application.ensure_all_started(&1))
  end
```

So then in the call I expose for command, I do something like:

```
def run_and_exit() do
   foo.ensure_all_deps_started
   foo.prepare_or_upgrade_stuff
   :init.stop
end
```

**Comment by martin-langhoff** [comment]
Formatting of complex command args is still terrible. Might have a patch to propose soon.

**Comment by jbsmith** [comment]
one additional item I had to use to get escript to run was that I had to set the ENV var for ERL_LIBS e.g.
 ERL_LIBS = /opt/app/rel/MYAPP/lib  /opt/app/rel/MYAPP/bin/MYAPP escript bin/my_release_tasks.escript args

**Comment by martin-langhoff** [comment]
Related #354 #359 for rpc  args parsing.

I am of the opinion that `rpc` and `command` should auto-format multiple args. For example

```
rpc module function arg1 arg2 arg3`

... should be transformed into

rpc module function '[<<"arg1">>,<<"arg2">>,<<"arg3">>].'
```

so that functions that take in parameters work similarly under Mix and exrm command invocations. To avoid mangling already-formatted inputs, it would only kick in if there's more than one parameter and the group does not end in '.'.

Admittedly, it is a bit of DWIM.

@bitwalker - thoughts?
